### PR TITLE
Support multiple elk instances in one process.

### PR DESCRIPTION
**Pull request by gjbadros**
This moves _message_handler from being file-local in messages.py to being a field in the Elk object.

Pipfile and Pipfile.lock changes seem spurious, but those weren't listed in .gitignore, so I left the diffs.

### CODE REVIEW

**Comment by gwww** [comment]
Thank you! Not what I was expecting. I was thinking you were talking about the Elkm1 library. No worries though. Since this is a rather substantial PR could you post on the discord channel (there's a long ElkM1 thread there) and let people know to see if they will try it out. I'd like to get more eyes on it than just mine, and, I won't be able to properly test this for a while.

**Comment by gwww** [comment]
I'll take a look.

**Comment by gjbadros** [comment]
It's both a change to the elkm1 library and a change to hass's elkm1
component.  The pr was for the elkm1 library.  Is that what you want posted
on discord?

It's actually not a big change in principle... Just touches all the files
with a minor change.

**Comment by gwww** [comment]
Overall looks pretty good. Thank you for the changes to support multiple Elk panels!

I was confused because the patch was for hass then the pr was for the lib. I now see you have changed both. My comment about discord is for the hass change, although letting people know that the lib is changing is also a good thing.

I think that the add_message_handler should be a method on Elk not on Element. The _message_handlers resides with Elk and should be manipulated from there. Does that sound right to you?

**Comment by gjbadros** [comment]
The methods on Element and Elements are just helper functions -- are you
saying to move message_handler out of the message.py?  I was trying to
avoid message.py taking a dependency on elk.py's Elk object, and as such,
add_message_handler staying in the code with the rest of the implementation
of message handling made some sense and was a more minimal change.

I initially considered just moving all of message.py into elk.py, but that
seemed like a bigger change so went with this simpler approach.

Greg

**Comment by gwww** [comment]
What I was suggesting is add `add_message_handler` method in `Elk` class.

I do believe is that the `message_handlers` array should be in `Elk`, as you have changed. 

What is causing me to wonder is having a "helper" in `Element`; it feels out of place. I'm OK with leaving `add_message_handler` in `message.py` as a helper. I view all of `message.py` to be a util/helper file.

Does that make sense?

**Comment by gjbadros** [comment]
How's this look now?  I kept the helpers in Element and Elements, but they're simply forwarding functions.  add_message_handler helper is still in message.py and elk.py has an _add_message_handler that calls that helper and is forwarded to by Element and Elements.

**Comment by gjbadros** [comment]
Just noticed a bug -- there's a couple more file globals in util.py -- stay tuned for an update to the PR. It was manifesting itself by not all the elk's updating their zone names (I had thought that was a bug in my hass implementation that I've not PR'd yet, but it's an omission in my library improvements)

**Comment by gwww** [comment]
Cool. Love what you're doing!

I think that I've figured out what was bugging me about the message
handling. Although I stand to still be wrong.

It's that the message handlers array is passed around and accessed in
multiple places.

Please be patient as we try to get it "right".

Currently thinking that the code in message.py should be a class. An
instance of the class is created by the main Elk class. This would
encapsulate all the message handler code. Perhaps the decoders too.

Does that make sense?

So you know, when I wrote the library, some parts were refactored a dozen
times until they felt right. 🤔

As you can see, I'm typing on my phone, so hard to be too much more
expressive at the moment.

Cheers, Glenn
______
Sent using an itsy bitsy keyboard. Expect typos and brevity.

**Comment by gjbadros** [comment]
Have another look please -- I've looked for other file-global objects and I'm pretty confident that the ones in util.py are the last that I missed.

And yea, I agree with you that putting some of the messages code in a class might make it a bit cleaner.  I can imagine a semantics-preserving rewrite that does that with a proper registration pattern (more like what you'd do in a static-typed language without using globals() and lexical creation of the function name to invoke) but that's a bigger rewrite that makes it less obvious that the change is correct.  It also might be unnecessary given how complete this code otherwise is.  My view is that with the refactoring to simplify that I implemented, the pay-off for encapsulating messages is not there and the importance of fixing the Elk singleton assumption is high.

Re: the test cases for get_descriptions, etc., I'm not sure there's a great way to do those and the test coverage of the class hierarchy is virtually zero, so now that that functionality lives inside Elk and Elements it is not out of place to not have good test coverage.  (That written, I've never used this Mock() class before so maybe I'm missing something about how it could be done painlessly.)

**Comment by gwww** [comment]
Take a look at the new-eh branch. It's not perfect (timeout and unknown messages not working properly). Thoughts?

**Comment by gjbadros** [comment]
Cool! I don't love the name "Message" for the class -- maybe
MessageProcessor (since Handler is used for something else).  Message makes
me think the instances will be, well, messages.

Is the intent for this to replace my CL or just provide an example of what
the "Message" class could look like?  If the former, there's still file
globals in util.py related to sync handlers and description processing that
need to be moved into elk.

If the latter, it might be nice to have a dictionary of all the allowed
commands to validate the command before trying to look it up.  I suppose
it's not too risky given the extra structure on the method name
(/_.._decode/), but I think getattr() chases up the inheritance tree so it
could be non obvious that there's a surprise method that could be invoked
by a nefarious Elk-pretending server.

Do you know why unknown messages and timeouts don't work?

**Comment by gwww** [comment]
Thanks for the feedback! Yah, `Message` is a bit weak. Can change. I like the way the class came together. It feels right, with the handlers stored in the Elk class and the the message.py nicely holds all the message stuff.

I know why timeout and unknown don't work. Just need more refactoring.

How about this, I add this change for the message handling, you submit a PR for the stuff in util.py?  Or... I'm open to suggestions! I wasn't really planning on getting into this but couldn't resist a nice design problem!

**Comment by gjbadros** [comment]
I probably can just pull down your message.py change and then almost
everything else in my fork will work... That would be easier for me to
manage from a change perspective. Or remake your message class changes on
my fork?  I think I gave you edit permission. Then do the pull into the
main repo.

**Comment by gwww** [comment]
I've changed a couple of things beyond the add_handler stuff. How about I merge what I have into master, you can pull the latest, then PR against that.

I've fixed unknown. Working on timeout. Won't be hard, might affect a couple of files (elk.py and proto.py)

**Comment by gjbadros** [comment]
Ok sounds good.

**Comment by gwww** [comment]
Master up to date with all fixes I know about; changed `Message` to `MessageDecode`

**Comment by gjbadros** [comment]
Ok, thx; I'll try this weekend to complete my second set of changes to
avoid the single-instance elk assumption.

File: elkm1_lib/elk.py [hunk 1]
```diff
@@ -25,9 +25,12 @@ def __init__(self, config, loop=None):
         self.connection_lost_callbk = None
         self._connection_retry_timer = 1
         self._message_decode = MessageDecode()
-
+        self._sync_handlers = []
+        self._descriptions_in_prog = {}
```
Comment (gwww) [inline elkm1_lib/elk.py]: Would be nicer to use full words. Use `_descriptions_in_progress` (`prog` is ambiguous)

File: elkm1_lib/elements.py [hunk 2]
```diff
@@ -88,9 +91,19 @@ def _got_desc(self, descriptions):
             if descriptions[element.index] is not None:
                 element.setattr('name', descriptions[element.index], True)
 
+    def _add_message_handler(self, message_type, handler):
```
Comment (gwww) [inline elkm1_lib/elements.py]: This appears to be defined twice in class. Can it be done away with altogether?

**Comment by gjbadros** [comment]
Make those changes; see the latest I just commited.  I dropped some testing
code for the newly-in-elk sync_handler and descriptions handling... if you
have ideas on how to get that testing back, I'd love for you to accept this
pull request and make those changes, then I can test more for my system off
of your master before you do a new release.

**Comment by gwww** [comment]
I bumped the version to 0.7.14; make a small tidy up

**Comment by gwww** [comment]
PyPi updated with 0.7.14 version

**Pull request merged**